src/poisson.f90

```diff
@@ -298,89 +298,123 @@ subroutine poisson_000(rhs)
 
     integer :: nx,ny,nz, i,j,k
 
-    complex(mytype) :: cx
-    real(mytype) :: rl, iy
-    external cx, rl, iy
+    complex(mytype) :: cw1_tmp
+    real(mytype) :: ntot
+#ifdef DEBUG
+    integer, dimension(3) :: dim3d
+#endif
 
     nx = nx_global
     ny = ny_global
     nz = nz_global
+    ntot = real(nx, kind=mytype) &
+         * real(ny, kind=mytype) &
+         * real(nz, kind=mytype)
 
     if (.not. fft_initialised) then
        call decomp_2d_fft_init(PHYSICAL_IN_Z)
        fft_initialised = .true.
     end if
 
     ! compute r2c transform
+#ifdef DEBUG
+    dim3d = shape(rhs)
+    write(*,*) 'RHS before transform'
+    do k = 1, dim3d(3),dim3d(3)/2+1
+      do j = 1, dim3d(2),dim3d(2)/2+1
+        do i = 1, dim3d(1),dim3d(1)/2+1
+          print "(i3,i3,i3,1x,e12.5,1x,e12.5)", i, j, k, rhs(i,j,k)
+        enddo
+      enddo
+    enddo
+#endif
     call decomp_2d_fft_3d(rhs,cw1)
+#ifdef DEBUG
+    dim3d = shape(cw1)
+    write(*,*) 'CW1 after transform'
+    do k = 1, dim3d(3),dim3d(3)/2+1
+      do j = 1, dim3d(2),dim3d(2)/2+1
+        do i = 1, dim3d(1),dim3d(1)/2+1
+          print "(i3,i3,i3,1x,e12.5,1x,e12.5)", i, j, k, real(cw1(i,j,k)),&
```

Comment:
The array cw1 is accessed with a local index (1:dim). However, in the loop below, the index is global (xst:xen). This should be fixed.

Reply:
I was just copying paste from another piece of code. I’ll modify this.